Allow clients to request most recent container logs

smarterclayton opened the pull request:
Many users attempt to use 'kubectl logs' in order to find the logs for a container, but receive no logs or an error telling them their container is not running. The fix in this case is to run with '--previous', but this does not match user expectations for the logs command (and many never discover --previous).

This change makes the default kubelet behavior to return logs for the most recent running or terminated container, consistent with docker logs.

Clean up error messages in the kubelet log path to be consistent and give
users a more predictable experience.

Fixes #4526 and #20256

smarterclayton committed:
Kubelet server was not returning a 500 on errors writing logs

Writing 200 first masks the second error. 200 is defaulted by the Go
http stack automatically.

smarterclayton commented:
Log error cleanup:

```
$ kubectl logs test
Error from server: container "test" in pod "test" is waiting to start - no logs yet
```

smarterclayton commented:
@k8s-bot test this

smarterclayton commented:
Needs @kubernetes/kube-api for suggestions for the flag or alternate ways of telling the kubelet to Do-The-Right-Thing in a backcompat way.

bgrant0607 commented:
We could probably automatically use the mostRecent behavior, except when follow is specified.

smarterclayton commented:
I'm fine with changing the old behavior.  It's not a very good behavior for
the vast majority of use cases.

On Thu, Jan 28, 2016 at 9:52 PM, Brian Grant <notifications@github.com>
wrote:

> We could probably automatically use the mostRecent behavior, except when
> follow is specified.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/20258#issuecomment-176544254>
> .
>

bgrant0607 commented:
Does follow automatically continue across container restarts?

smarterclayton commented:
Not today, it breaks when the container dies.

On Thu, Jan 28, 2016 at 10:01 PM, Brian Grant <notifications@github.com>
wrote:

> Does follow automatically continue across container restarts?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/20258#issuecomment-176545753>
> .
>

smarterclayton commented:
So there's no actual way to tell the difference from a client from doing a
?follow against a container that exits while you are downloading the
initial log snippet and a container that terminated a while back other than
timestamps.

On Thu, Jan 28, 2016 at 10:02 PM, Clayton Coleman <ccoleman@redhat.com>
wrote:

> Not today, it breaks when the container dies.
>
> On Thu, Jan 28, 2016 at 10:01 PM, Brian Grant <notifications@github.com>
> wrote:
>
>> Does follow automatically continue across container restarts?
>>
>> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/kubernetes/kubernetes/pull/20258#issuecomment-176545753>
>> .
>>
>
>

smarterclayton committed:
Allow clients to request most recent container logs

Many users attempt to use 'kubectl logs' in order to find the logs
for a container, but receive no logs or an error telling them their
container is not running. The fix in this case is to run with '--previous',
but  this does not match user expectations for the logs command.

This commit changes the behavior of the Kubelet to return the logs of
the currently running container or the previous running container unless
the user provides the "previous" flag. If the user specifies "follow"
the logs of the most recent container will be displayed, and if it is
a terminated container the logs will come to an end (the user can
repeatedly invoke 'kubectl logs --follow' and see the same output).

Clean up error messages in the kubelet log path to be consistent and
give users a more predictable experience.

Have the Kubelet return 400 on invalid requests

smarterclayton commented:
I simplified this down to just honoring "previous" but making the default
behavior "get the most recent container".  I did not add anything special
for follow - the behavior now is consistent with docker (a terminated
container can be invoked with docker logs --follow NAME and simply displays
the container logs).

On Thu, Jan 28, 2016 at 10:03 PM, Clayton Coleman <ccoleman@redhat.com>
wrote:

> So there's no actual way to tell the difference from a client from doing a
> ?follow against a container that exits while you are downloading the
> initial log snippet and a container that terminated a while back other than
> timestamps.
>
> On Thu, Jan 28, 2016 at 10:02 PM, Clayton Coleman <ccoleman@redhat.com>
> wrote:
>
>> Not today, it breaks when the container dies.
>>
>> On Thu, Jan 28, 2016 at 10:01 PM, Brian Grant <notifications@github.com>
>> wrote:
>>
>>> Does follow automatically continue across container restarts?
>>>
>>> —
>>> Reply to this email directly or view it on GitHub
>>> <https://github.com/kubernetes/kubernetes/pull/20258#issuecomment-176545753>
>>> .
>>>
>>
>>
>

kargakis commented on pkg/kubelet/kubelet.go:
```diff
@@ -2508,15 +2505,16 @@ func (kl *Kubelet) GetKubeletContainerLogs(podFullName, containerName string, lo
+	switch podStatus.Phase {
+	case api.PodRunning, api.PodSucceeded, api.PodFailed:
+	default:
+		return fmt.Errorf("pod %q is in an unknown state and logs are not available", pod.Name)
```
new and pending are not unknown states

smarterclayton replied:
Same code as before, but agree the message should be changed.  I briefly debated having the kubelet hold the request until the pod became available (because in practice most clients want to do something like that) but it would require a lot of extra timeout logic that we can add in the future.

smarterclayton replied:
After looking at this some more, I'm going to remove this check.  If the pod can reference the container, and the container exists, we should show logs.  There's no point to excessively gating which logs we can return.

smarterclayton replied:
Ok, I went back and restructured this more, please review again.

1. Phase is not considered - if we can find a container, great, otherwise, we already return an error to the client
2. I've added more error information when a container can't be logged - two most common scenarios for end users are going to be something like "can't pull image" or "image pull backoff", so I explicitly return an error message for those states, and then otherwise print the reason that the container is in waiting state (when there is no previous container).
3. Simplified some of the code to be more readable.

kargakis commented:
one comment, lgtm otherwise!

kargakis commented on pkg/kubelet/kubelet.go:
```diff
@@ -2449,32 +2449,51 @@ func (kl *Kubelet) LatestLoopEntryTime() time.Time {
+
+	switch {
+	case previous:
+		if lastState.Terminated == nil {
```
Is it possible that this is still running?

smarterclayton replied:
Unless the rest of the Kubelet is broken, no.  Kubelet doesn't set lastState.Terminated until the container is observed killed and stopped.

kargakis replied:
ok

kargakis commented:
LGTM
